opm/upscaling/RelPermUtils.hpp

```diff
@@ -0,0 +1,161 @@
+  double getVoigtValue(const SinglePhaseUpscaler::permtensor_t& K, int voigt_idx);
+
+  //! \brief Set value in tensor
+  //! \param[out] K The tensor to set value in
+  //! \param[in] voigt_idx The voigt index for value to set (0..8)
+  //! \param[in] val Value to set in tensor
+  void setVoigtValue(SinglePhaseUpscaler::permtensor_t& K, int voigt_idx, double val);
+
+  //! \brief Helper class for relperm upscaling applications.
+  class RelPermUpscaleHelper {
+    public:
+      bool isMaster;                               //!< Whether this is the master MPI node or not.
+      bool doEclipseCheck;                         //!< Whether to check that input relperm curves include relperm at critical saturation points.
+      bool anisotropic_input;                      //!< Whether input eclipse file has diagonal anisotrophy.
+      int points;                                  //!< Number of saturation points to upscale for.
+      int tensorElementCount;                      //!< Number of independent elements in resulting tensor.
+      bool upscaleBothPhases;                      //!< Whether to upscale both phases.
+      std::vector<double> WaterSaturation;         //!< Re-upscaled water saturation for the computed pressure points.
+      SinglePhaseUpscaler::permtensor_t permTensor; //!< Tensor of upscaled results.
+      std::vector<int> satnums;                    //!< Cell satnums.
+      std::vector<MonotCubicInterpolator> InvJfunctions; //!< Inverse of the loaded J-functions.
+      //! \brief Relperm-curves for each (component->phase->stone type)
+      std::array<std::array<std::vector<MonotCubicInterpolator>,2>,3> Krfunctions;
+      SinglePhaseUpscaler::BoundaryConditionType boundaryCondition; //!< Boundary conditions to use.
+      std::vector<MonotCubicInterpolator> SwPcfunctions; //!< Holds Sw(Pc) for each rocktype.
+      std::string saturationstring; //!< Fluid system type.
+      size_t tesselatedCells; //!< Number of "active" cells (Sintef interpretation of "active")
+      double volume; //!< Total volume.
+      double poreVolume; //!< Total pore volume.
+      std::vector<double> pressurePoints; //!< Vector of capillary pressure points between Swor and Swir.
+
+      //! \brief Default constructor.
+      //! \param[in] mpi_rank Rank of this process (for parallel simulations).
+      //! \param[in] options Options structure.
+      //! \details Uses the following options: fluids
+      RelPermUpscaleHelper(int mpi_rank, std::map<std::string,std::string>& options_);
+
+      //! \brief Collect results from all MPI nodes.
+      void collectResults();
+
+      //! \brief Calculate relperm values from phase permeabilities.
+      //! \param[in] phase The phase to calculate values for (0-indexed).
+      //! \return The phase permeability values.
+      std::vector<std::vector<double>> getRelPerm(int phase) const;
+
+      //! \brief Upscale single phase permeability
+      void upscaleSinglePhasePermeability();
```

Comment:
i have basically kept the blocks from the original programs,
and stuck them in functions corresponding to the block comments in the 
main programs -> see the main code.